pkg/scheduler/objects/queue.go

```diff
@@ -1292,3 +1306,71 @@ func (sq *Queue) String() string {
 	return fmt.Sprintf("{QueuePath: %s, State: %s, StateTime: %x, MaxResource: %s}",
 		sq.QueuePath, sq.stateMachine.Current(), sq.stateTime, sq.maxResource)
 }
+
+func (sq *Queue) incRunningApps() {
+	if sq == nil {
+		return
+	}
+	if sq.parent != nil {
+		sq.parent.incRunningApps()
+	}
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps++
+}
+
+func (sq *Queue) decRunningApps() {
+	if sq == nil {
+		return
+	}
+	if sq.parent != nil {
+		sq.parent.decRunningApps()
+	}
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps--
+}
+
+func (sq *Queue) incAllocatingAcceptedAppsIfCanRun(sa *Application) bool {
+	if sq == nil {
+		return false
+	}
+	success := true
+	if sq.parent != nil {
+		success = sq.parent.incAllocatingAcceptedAppsIfCanRun(sa)
+	}
+	return sq.InternalIncAllocatingAcceptedAppsIfCanRun(success, sa)
+}
+
+func (sq *Queue) InternalIncAllocatingAcceptedAppsIfCanRun(success bool, sa *Application) bool {
```

Comment:
This can be combined into incAllocatingAppsIfCanRun() as well, since it is a tail call. I think that improves the visibility.

Reply:
Sure, combined.